executor/aggfuncs/func_first_row.go

```diff
@@ -0,0 +1,474 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aggfuncs
+
+import (
+	"fmt"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/types/json"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+type basePartialResult4FirstRow struct {
+	// isNull indicates whether the first row is null.
+	isNull bool
+	// gotFirstRow indicates whether the first row has been got,
+	// if so, we would avoid evaluating the values of the remained rows.
+	gotFirstRow bool
+}
+
+type partialResult4FirstRowInt struct {
+	basePartialResult4FirstRow
+
+	val int64
+}
+
+type partialResult4FirstRowUint struct {
+	basePartialResult4FirstRow
+
+	val uint64
+}
+
+type partialResult4FirstRowFloat32 struct {
+	basePartialResult4FirstRow
+
+	val float32
+}
+
+type partialResult4FirstRowDecimal struct {
+	basePartialResult4FirstRow
+
+	val types.MyDecimal
+}
+
+type partialResult4FirstRowFloat64 struct {
+	basePartialResult4FirstRow
+
+	val float64
+}
+
+type partialResult4FirstRowString struct {
+	basePartialResult4FirstRow
+
+	val string
+}
+
+type partialResult4FirstRowTime struct {
+	basePartialResult4FirstRow
+
+	val types.Time
+}
+
+type partialResult4FirstRowDuration struct {
+	basePartialResult4FirstRow
+
+	val types.Duration
+}
+
+type partialResult4FirstRowJSON struct {
+	basePartialResult4FirstRow
+
+	val json.BinaryJSON
+}
+
+type firstRow4Int struct {
+	baseAggFunc
+}
+
+func (e *firstRow4Int) AllocPartialResult() PartialResult {
+	return PartialResult(new(partialResult4FirstRowInt))
+}
+
+func (e *firstRow4Int) ResetPartialResult(pr PartialResult) {
+	p := (*partialResult4FirstRowInt)(pr)
+	p.val, p.isNull, p.gotFirstRow = 0, false, false
+}
+
+func (e *firstRow4Int) UpdatePartialResult(sctx sessionctx.Context, rowsInGroup []chunk.Row, pr PartialResult) error {
+	p := (*partialResult4FirstRowInt)(pr)
+	if p.gotFirstRow {
+		return nil
+	}
+	for _, row := range rowsInGroup {
+		input, isNull, err := e.args[0].EvalInt(sctx, row)
+		if err != nil {
+			return errors.Trace(err)
+		}
```

Comment:
set `p.gotFirstRow = true` here, break the calculation after updating the partial result.